python/ray/dataframe/dataframe.py

```diff
@@ -4464,9 +4464,121 @@ def remote_func(df):
+                DataFrame.
+            other: A value or DataFrame of values to use for setting this.
+            inplace: Whether or not to operate inplace.
+            axis: The axis to apply over. Only valid when a Series is passed
+                as other.
+            level: The MultiLevel index level to apply over.
+            errors: Whether or not to raise errors. Does nothing in Pandas.
+            try_cast: Try to cast the result back to the input type.
+            raise_on_error: Whether to raise invalid datatypes (deprecated).
+
+        Returns:
+            A new DataFrame with the replaced values.
+        """
+
+        inplace = validate_bool_kwarg(inplace, 'inplace')
+
+        if isinstance(other, pd.Series) and axis is None:
+            raise ValueError("Must specify axis=0 or 1")
+
+        if level is not None:
+            raise NotImplementedError("Multilevel Index not yet supported on "
+                                      "Pandas on Ray.")
+
+        axis = pd.DataFrame()._get_axis_number(axis) if axis is not None else 0
+
+        cond = cond(self) if callable(cond) else cond
+
+        if not isinstance(cond, DataFrame):
+            if not hasattr(cond, 'shape'):
+                cond = np.asanyarray(cond)
+            if cond.shape != self.shape:
+                raise ValueError("Array conditional must be same shape as "
+                                 "self")
+            cond = DataFrame(cond, index=self.index, columns=self.columns)
+
+        zipped_partitions = self._copartition(cond, self.index)
+        args = (False, axis, level, errors, try_cast, raise_on_error)
+
+        @ray.remote
+        def where_helper(left, cond, other, *args):
+
+            left = pd.concat(ray.get(left.tolist()), axis=1)
+            # We have to reset the index and columns here because we are coming
+            # from blocks and the axes are set according to the blocks. We have
+            # already correctly copartitioned everything, so there's no
+            # correctness problems with doing this.
+            left.reset_index(inplace=True, drop=True)
+            left.columns = pd.RangeIndex(len(left.columns))
+
+            cond = pd.concat(ray.get(cond.tolist()), axis=1)
+            cond.reset_index(inplace=True, drop=True)
+            cond.columns = pd.RangeIndex(len(cond.columns))
+
+            if isinstance(other, np.ndarray):
+                other = pd.concat(ray.get(other.tolist()), axis=1)
+                other.reset_index(inplace=True, drop=True)
+                other.columns = pd.RangeIndex(len(other.columns))
+
+            return left.where(cond, other, *args)
+
+        if isinstance(other, DataFrame):
+            other_zipped = (v for k, v in self._copartition(other,
+                                                            self.index))
+
+            new_partitions = [where_helper.remote(k, v, next(other_zipped),
```

Comment:
Can k, v be converted to lists and passed in by reference? Ray will automatically deserialize then.

Reply:
not without merging them together, then also passing the length of the left. Performance-wise it's not much different.